frigate/monitoring/prometheus.py

```diff
@@ -0,0 +1,296 @@
+import json
+import re
+from urllib.request import urlopen
+from prometheus_client import CollectorRegistry
+from prometheus_client.metrics_core import GaugeMetricFamily, InfoMetricFamily, Metric
+from typing import Dict, Any, Generator
+
+
+def setupRegistry() -> CollectorRegistry:
+    myregistry = CollectorRegistry()
+    myregistry.register(CustomCollector())
+    return myregistry
+
+
+def add_metric(
+    metric: GaugeMetricFamily,
+    label: str,
+    stats: Dict[str, Any],
+    key: str,
+    multiplier: float = 1.0,
+) -> None:
+    try:
+        string = str(stats[key])
+        value = float(re.findall(r"\d+", string)[0])
+        metric.add_metric([label], value * multiplier)
+    except (KeyError, TypeError, IndexError):
+        pass
+
+
+class CustomCollector(CollectorRegistry):
+    def __init__(self) -> None:
+        self.stats_url = "http://localhost:5000/api/stats"
+        self.process_stats: Dict[str, Dict[str, float]] = {}  # Add a type hint here
+
+    def add_metric_process(
+        self,
+        metric: GaugeMetricFamily,
+        stats: Dict[str, Any],
+        camera_name: str,
+        pid_name: str,
+        process_name: str,
+        cpu_or_memory: str,
+        process_type: str,
+    ) -> None:
+        try:
+            pid = str(stats[camera_name][pid_name])
+            label_values = [pid, camera_name, process_name, process_type]
+            metric.add_metric(label_values, self.process_stats[pid][cpu_or_memory])
+            del self.process_stats[pid][cpu_or_memory]
+        except (KeyError, TypeError, IndexError):
+            pass
+
+    def collect(self) -> Generator[Metric, None, None]:
+        stats = json.loads(urlopen(self.stats_url).read())
```

Comment:
I might be missing something - but why do a request to the Frigate API rather than importing what we need?

Reply:
I'm also curious about this. Any reason we need to use the http endpoint here?

Reply:
Its been a minute, I am trying to recollect.... I believe I couldn't find a way to pass the raw stats due to the life cycle of the Prometheus client. When the custom collector gets passed to the client start up we have this code 

![image](https://github.com/blakeblackshear/frigate/assets/6330506/3b18ddd1-7eb6-428a-9898-7bc852ee9f4d)

Which will trigger

![image](https://github.com/blakeblackshear/frigate/assets/6330506/b25d6348-3588-4a6a-9f5d-f4096bbf7c2b)

When the custom collector is registered, we then face

![image](https://github.com/blakeblackshear/frigate/assets/6330506/a25397b1-70ad-4431-931f-410ac304f84b)

So now custom collector is called, thus "stats = json.loads(urlopen(self.stats_url).read())" is called


I remember fighting with this for days because I needed a way to bootstrap the Prometheus client with out it crashing. My mind is fuzzy here now, but there was for sure a issue with first time load and crashing. I recall reading all the tutorials and scanning GitHub to find examples and all of them did it this way. 

If have suggestions I'm all ears. If someone wants to open a PR against mine, Im welcome to that too.

Reply:
The use of the http stats is from my docker container. But this doesn't need to use the API to get the stats data. It could alternatively use a similar method to collect the stats as like how the API stats are generated (I haven't investigated this).

Reply:
> The use of the http stats is from my docker container. But this doesn't need to use the API to get the stats data. It could alternatively use a similar method to collect the stats as like how the API stats are generated (I haven't investigated this).

Unable to get this to work. Only way I've gotten it to work is with the http route, for the same reasons I posted above. Im either missing something extremely easy here or the app is not architected to fit the Prometheus exporter library.